caffe2/contrib/nccl/cuda_nccl_gpu.cc

```diff
@@ -126,19 +126,23 @@ void runNCCL(const NCCLExecution& ex, F&& f) {
     CUDA_CHECK(cudaEventRecord(context->master_event_, ex.stream));
   }
 
-  for (auto i = 0; i < ex.elements.size(); ++i) {
-    auto& ctx = ex.elements[i];
-    DeviceGuard g(ctx.device);
-    auto& comm = comms[i];
-    auto& stream = streams[i];
-    auto& event = events[i];
-
-    DCHECK_EQ(ctx.device, GetGPUIDForPointer(ctx.src->raw_data()));
-    CUDA_CHECK(cudaStreamWaitEvent(stream, context->master_event_, 0));
-    f(ctx, comm, stream);
-    // Record an event on each children stream that we have finished
-    // our computation
-    CUDA_CHECK(cudaEventRecord(event, stream));
+  {
+    // lock out alloc / free while NCCL launches
+    std::lock_guard<std::mutex> lock(CUDAContext::mutex());
```

Comment:
I guess read/write lock would be better, so that NCCL could run in parallel? But of course currently we don't run NCCL in parallel (in data_parallel_model).